### PR TITLE
Media transfer integration

### PR DESCRIPTION
**Pull request by bnookala**
## Purpose

Introduces file preview to chat window. Example: 

Screen Shot 2021-01-25 at 10 35 21 AM

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Upload a file, verify it shows up in the chat window.

## What to Check
Verify that the following are valid
* That an image preview renders in the chat window.

## Other Information

Chat renderer intercepts messages and renders them as image previews if they contain a certain json payload. See: https://github.com/navzam/communication-services-web-calling-hero/compare/main...bnookala:media-transfer-integration?expand=1#diff-f706cefd0ab650077677cc7c3f9e111eef5594b8d74189fd63d6f0dec944e7f5R310 for more information on how this works.

### CODE REVIEW

File: Calling/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (edaena) [inline Calling/.vscode/launch.json]: Is this file needed?

File: Calling/ClientApp/src/components/ChatThread.tsx [hunk 2]
```diff
@@ -302,6 +307,22 @@ export default (props: ChatThreadProps): JSX.Element => {
...
+  const isUploadedFileMessage = (messageContent: string): Boolean => {
+    let messageContentJson = undefined;
+    try {
+      messageContentJson = JSON.parse(messageContent);
+    } catch (error) {
+      // Not a file upload event.
+      return false;
+    }
+
+    if (messageContentJson.event && messageContentJson.event !== "FileUpload") {
```
Comment (edaena) [inline Calling/ClientApp/src/components/ChatThread.tsx]: nit: can also just return the conditional:
return messageContentJson.event && messageContentJson.event == "FileUpload"

File: Calling/ClientApp/src/core/sideEffects.ts [hunk 3]
```diff
@@ -447,34 +447,45 @@ export const getFile = (fileId: string) => {
...
+
 export const uploadSelectedFile = (file: File) => {
   return async (dispatch: Dispatch, getState: () => State) => {
     const state = getState();
     const userId = state.sdk.userId;
-    if (userId === undefined) {
+    const threadId = state.thread.threadId;
+
+    if (userId === undefined || threadId === undefined) {
       console.error(`Failed to make sendFile() API call because userId is undefined`);
```
Comment (edaena) [inline Calling/ClientApp/src/core/sideEffects.ts]: or threadId

File: Calling/NodeApi/index.ts [hunk 4]
```diff
@@ -237,19 +248,23 @@ app.get('/getEnvironmentUrl', fakeAuthMiddleware, async (req, res) => {
...
 
-    let chatThreadClient= await chatClient.createChatThread(createThreadRequest);
+    let chatThreadClient = await chatClient.createChatThread(createThreadRequest);
     let threadId = chatThreadClient.threadId;
     tokenStore[threadId] = new CommunicationUserToken(threadId, tokenResponse.user.communicationUserId, tokenResponse.token);
 
     return res.status(200).send(threadId);
  });
 
+ const generateNewModeratorAndThread = async () => {
```
Comment (edaena) [inline Calling/NodeApi/index.ts]: Try to pull the latest, this should have been removed.

**Review by edaena — CHANGES_REQUESTED** [review]

File: Calling/ClientApp/src/components/ChatThread.tsx [hunk 5]
```diff
@@ -302,6 +307,22 @@ export default (props: ChatThreadProps): JSX.Element => {
     setMessagesWithAttachedRef(newMessagesWithAttached);
   };
 
+  const isUploadedFileMessage = (messageContent: string): Boolean => {
```
Comment (navzam) [inline Calling/ClientApp/src/components/ChatThread.tsx]: Instead of a `Boolean`, if this returned the parsed JSON (maybe a `FileMessage` object or `null`), then you could avoid re-parsing it later.

File: Calling/ClientApp/src/components/ChatThread.tsx (hunk 2 above)
Comment (navzam) [inline Calling/ClientApp/src/components/ChatThread.tsx]: Should the condition be `!messageContentJson.event || messageContentJson.event !== "FileUpload"`? Otherwise, a valid JSON without an `event` property would return `true`.

File: Calling/ClientApp/src/components/ChatThread.tsx [hunk 6]
```diff
@@ -322,6 +343,30 @@ export default (props: ChatThreadProps): JSX.Element => {
...
               items={messagesWithAttached.map((message: any, index: number) => {
+                if (isUploadedFileMessage(message.content)) {
+                  const messageContent = JSON.parse(message.content);
+                  if (!messageContent) {
+                    return {};
+                  }
+
+                  return {
+                    key: index,
+                    contentPosition: message.mine ? 'end' : 'start',
```
Comment (navzam) [inline Calling/ClientApp/src/components/ChatThread.tsx:352]: Should the file card be rendering on the left/right depending on who sent it? When I run it, file messages are always rendered on the left side of the chat.

File: Calling/ClientApp/src/components/ChatThread.tsx [hunk 7]
```diff
@@ -322,6 +343,30 @@ export default (props: ChatThreadProps): JSX.Element => {
...
+                    message: (
+                      <Card aria-roledescription="image card">
```
Comment (navzam) [inline Calling/ClientApp/src/components/ChatThread.tsx:354]: Nice to have: Show the display name of the user who sent the file. I'm working on storing ACS user IDs in Storage, which I'm guessing would be required to achieve this, so this can definitely go on the backlog.

File: Calling/ClientApp/src/containers/GroupCall.ts [hunk 8]
```diff
@@ -31,6 +31,7 @@ const mapStateToProps = (state: State, props: GroupCallProps) => ({
   },
   remoteParticipants: state.calls.remoteParticipants,
   streams: state.streams.streams,
+  threadId: state.thread.threadId,
```
Comment (navzam) [inline Calling/ClientApp/src/containers/GroupCall.ts]: Is this used anywhere in `GroupCall`?

File: Calling/NodeApi/index.ts [hunk 9]
```diff
@@ -56,8 +54,8 @@ const [
 });
 
 const blobContainerName = 'files';
-const fileMetadataTableName = 'fileMetadata';
-const userDetailTableName = 'userDetails';
+const tableName = 'fileMetadata';
```
Comment (navzam) [inline Calling/NodeApi/index.ts]: Might need to pull? These names were made more specific since we're adding more tables

**Comment by navzam** [comment]
Now that file events are in the chat thread, we should change file refresh from polling to event-based. Should we add a separate user story to cover that?

File: Calling/.vscode/launch.json (hunk 1 above)
Comment (bnookala) [reply Calling/.vscode/launch.json]: don't think so, will remove

File: Calling/ClientApp/src/components/ChatThread.tsx (hunk 2 above)
Comment (bnookala) [reply Calling/ClientApp/src/components/ChatThread.tsx]: fixed by returning the opposite of this condition

File: Calling/ClientApp/src/components/ChatThread.tsx (hunk 6 above)
Comment (bnookala) [reply Calling/ClientApp/src/components/ChatThread.tsx:352]: yeah, i noticed this too - might be because there is no "mine" property on this message - perhaps we can add this is a backlog item?

File: Calling/ClientApp/src/components/ChatThread.tsx (hunk 7 above)
Comment (bnookala) [reply Calling/ClientApp/src/components/ChatThread.tsx:354]: sounds good, i'll create a follow up ticket

File: Calling/ClientApp/src/containers/GroupCall.ts (hunk 8 above)
Comment (bnookala) [reply Calling/ClientApp/src/containers/GroupCall.ts]: oops, might have been left in while trying to debug something else.

File: Calling/NodeApi/index.ts (hunk 9 above)
Comment (bnookala) [reply Calling/NodeApi/index.ts]: fixed

File: Calling/ClientApp/src/components/ChatThread.tsx (hunk 5 above)
Comment (bnookala) [reply Calling/ClientApp/src/components/ChatThread.tsx]: fixed

**Comment by bnookala** [comment]
@navzam yes let's add a user story to cover event based updates :)

**Comment by navzam** [comment]
> @navzam yes let's add a user story to cover event based updates :)

 @bnookala Added

**Review by navzam — APPROVED** [review]

**Comment by evanlouie** [comment]
lgtm 👍

**Review by edaena — APPROVED** [review]
